pkg/tfbridge/detailed_diff.go

```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
 	return keysSlice
 }
 
+func isTypeShapeMismatched(val resource.PropertyValue, propType shim.ValueType) bool {
+	if !isPresent(val) {
+		return false
+	}
+	switch propType {
+	case shim.TypeList:
+		return !val.IsArray()
+	case shim.TypeSet:
+		return !val.IsArray()
+	case shim.TypeMap:
+		return !val.IsObject()
+	default:
+		return false
+	}
+}
+
+func lookupSchemas(
+	path propertyPath, tfs shim.SchemaMap, ps map[string]*info.Schema,
+) (shim.Schema, *info.Schema, error) {
+	schemaPath := PropertyPathToSchemaPath(resource.PropertyPath(path), tfs, ps)
+	return LookupSchemas(schemaPath, tfs, ps)
+}
+
+// walkPropertyValue walks a property value and calls the visitor function for each path in the property value.
+func walkPropertyValue(
+	val resource.PropertyValue, path propertyPath, visitor func(propertyPath, resource.PropertyValue) bool,
+) bool {
+	if !visitor(path, val) {
+		return false
+	}
+
```

Comment:
Again, I'd like to see an assert that we are not hitting marker types.

Reply:
Just was commenting on another PR, but this is not a complete walker on PropertyValue.  

E.G.

https://github.com/pulumi/pulumi-terraform-bridge/blob/master/unstable/propertyvalue/propertyvalue.go#L42

If we had a complete walker on property value we would not need to double-test it. I know Go is about "a bit more duplication is better than a bit more sharing", but if we can reuse that makes coverage goals easier to achieve.